internal/app/orchestrator/orchestrator.go

```diff
@@ -90,16 +105,172 @@ func New(ctx context.Context, l log.Logger, mapper mapper.Mapper, upstreamClient
 //
 // Cancel is an optional function to release resources in the producer. If
 // provided, the consumer may call this function multiple times.
-func (c *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
-	// TODO implement.
-	return nil, nil
+func (o *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
+	ctx := context.Background()
+
+	o.downstreamResponseChannelsMu.Lock()
+	if o.downstreamResponseChannels[&req] == nil {
+		// If this is the first time we're seeing the request from the
+		// downstream client, initialize a channel to feed future responses.
+		o.downstreamResponseChannels[&req] = make(chan gcp.Response)
+	}
+	o.downstreamResponseChannelsMu.Unlock()
+
+	aggregatedKey, err := o.mapper.GetKey(req)
+	if err != nil {
+		// Can't map the request to an aggregated key. Log and continue to
+		// propagate the response upstream without aggregation.
+		o.logger.With("err", err).With("req", req).Error(ctx, "failed to get aggregated key")
+		// Mimic the aggregated key.
+		// TODO (insert issue) This key needs to be made more granular to
+		//      uniquely identify a request. User-specified defaults?
+		aggregatedKey = fmt.Sprintf("%s%s_%s", unaggregatedPrefix, req.GetNode().GetId(), req.GetTypeUrl())
+	}
+
+	// Register the watch for future responses.
+	err = o.cache.AddRequest(aggregatedKey, req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req", req).Error(ctx, "failed to add watch")
+		close(o.downstreamResponseChannels[&req])
+		// Clean up.
+		o.downstreamResponseChannelsMu.Lock()
+		delete(o.downstreamResponseChannels, &req)
+		o.downstreamResponseChannelsMu.Unlock()
+		return nil, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil {
+		// If we have a cached response, immediately return the result.
+		o.downstreamResponseChannels[&req] <- gcp.Response{
```

Comment:
im thinking about this scenario:
new sidecar -> value available in cache -> send immediate response

response gets applied on sidecar -> sidecar asks for another update 

At this point, we have to send back a channel which has nothing, so that it can be fulfilled later. 
Can you describe briefly the code path that enables the scenario?

Reply:
Please take a look again and see if the workflow makes more sense.

new sidecar -> value in cache -> immediate response -> watch is still opened if it doesn't already exist

new response from upstream -> fan out to watches -> channel gets populated.